src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDiagnosticFactory.cs

```diff
@@ -374,6 +374,53 @@ public static RazorDiagnostic CreateBindAttribute_FormatNode_Unsupported(SourceS
                 BindAttribute_FormatNode_Unsupported,
                 source ?? SourceSpan.Undefined);
             return diagnostic;
+
+        public static readonly RazorDiagnosticDescriptor DuplicateMarkupAttribute =
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}10007",
+                () => "The attribute '{0}' is used two or more times for this element. Attributes must be unique (case-insensitive).",
+                RazorDiagnosticSeverity.Error);
+
+        public static RazorDiagnostic Create_DuplicateMarkupAttribute(string attributeName, SourceSpan? source = null)
+        {
+            return RazorDiagnostic.Create(DuplicateMarkupAttribute, source ?? SourceSpan.Undefined, attributeName);
+        }
+
+        public static readonly RazorDiagnosticDescriptor DuplicateMarkupAttributeDirective =
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}10008",
+                () =>
+                    "The attribute '{0}' is used two or more times for this element. Attributes must be unique (case-insensitive). " +
+                    "The attribute '{0}' is used by the '{1}' directive attribute.",
+                RazorDiagnosticSeverity.Error);
+
+        public static RazorDiagnostic Create_DuplicateMarkupAttributeDirective(string attributeName, string directiveAttributeName, SourceSpan? source = null)
+        {
+            return RazorDiagnostic.Create(DuplicateMarkupAttributeDirective, source ?? SourceSpan.Undefined, attributeName, directiveAttributeName);
+        }
+
+        public static readonly RazorDiagnosticDescriptor DuplicateComponentParameter =
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}10009",
+                () => "The component parameter '{0}' is used two or more times for this component. Parameters must be unique (case-insensitive).",
+                RazorDiagnosticSeverity.Error);
+
+        public static RazorDiagnostic Create_DuplicateComponentParameter(string attributeName, SourceSpan? source = null)
+        {
+            return RazorDiagnostic.Create(DuplicateComponentParameter, source ?? SourceSpan.Undefined, attributeName);
+        }
+
+        public static readonly RazorDiagnosticDescriptor DuplicateComponentParameterDirective =
+            new RazorDiagnosticDescriptor(
+                $"{DiagnosticPrefix}10010",
+                () =>
+                    "The component parameter '{0}' is used two or more times for this component. Parameters must be unique (case-insensitive). " +
```

Comment:
> Parameters must be unique (case-insensitive).

I thought you said parameters are [case **sensitive**](https://github.com/aspnet/AspNetCore-Tooling/pull/597#discussion_r285736528)?

Reply:
If we're going to be using the terminology of "parameter" for `bind-ABC:XYZ`'s `ABC` @ajaybhargavb we may want to consider renaming the directive attribute "parameters" (`XYZ`) to be "modifiers" to avoid confusion.

Reply:
matching is case-sensitive. Duplication is case-insensitive. As in, you can't have parameters that differ only by case.

Reply:
Hmm that is a little confusing. What is the reason for not having matching be case-insensitive as well?